### PR TITLE
Miiverse and Error applet patch

### PR DESCRIPTION
**Pull request by shutterbug2000**
- Add non-NAND Miiverse patch
- Custom ban error, pulled from the Account server

Draft PR for now b/c I'd like it to be looked at a bit first :p

### CODE REVIEW

File: Makefile [hunk 1]
```diff
@@ -40,7 +40,7 @@ LDFLAGS	=	-g $(ARCH) $(RPXSPECS) -Wl,-Map,$(notdir $*.map)
 
 LDFLAGS	+=	-T$(WUMS_ROOT)/share/libkernel.ld $(WUPSSPECS)
 
-LIBS	:= -lwups -lmocha -lkernel -lwut
+LIBS	:= -lwups -lmocha -lkernel -lsdutils -lcurl -lz -lmbedtls -lmbedcrypto -lmbedx509 -lwut
```
Comment (ashquarky) [inline Makefile:43]: This build goes from 47K to 778K. I know why you've added curl but I do wonder if it's the best answer to the ban reason issue? Assuming we care about binary size, I've never gotten a straight answer from maschell as to how big the heap is other than "plenty"

File: src/main.cpp [hunk 2]
```diff
@@ -29,6 +29,17 @@
 #include "wut_extra.h"
 #include <utils/logger.h>
 #include "url_patches.h"
+#include <coreinit/filesystem.h>
+#include <cstring>
+#include <string>
+#include <sdutils/sdutils.h>
```
Comment (ashquarky) [inline src/main.cpp:35]: sdutils is seemingly unused?

File: src/main.cpp [hunk 3]
```diff
@@ -58,6 +69,19 @@ OSDynLoad_IsModuleLoaded(char const *name,
...
+const char wave_original[] = {
+	0x68, 0x74, 0x74, 0x70, 0x73, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00,
+	0x00, 0x00, 0x00, 0x00, 0x2E, 0x6E, 0x69, 0x6E, 0x74, 0x65, 0x6E, 0x64,
+	0x6F, 0x2E, 0x6E, 0x65, 0x74
+};
+const char wave_new[] = {
+	0x68, 0x74, 0x74, 0x70, 0x73, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00,
+	0x00, 0x00, 0x00, 0x00, 0x2E, 0x70, 0x72, 0x65, 0x74, 0x65, 0x6E, 0x64,
+	0x6F, 0x2E, 0x63, 0x63, 0x00
+};
```
Comment (ashquarky) [inline src/main.cpp:81]: aren't these just strings? You're not changing the bit before the nulls

File: src/main.cpp [hunk 4]
```diff
@@ -240,6 +302,17 @@ bool replace(uint32_t start, uint32_t size, const char* original_val, size_t ori
     return false;
 }
 
+uint32_t find(uint32_t start, uint32_t size, const char* original_val, size_t original_val_sz) {
```
Comment (ashquarky) [inline src/main.cpp:305]: Unused

File: src/main.cpp [hunk 5]
```diff
@@ -275,3 +348,101 @@ ON_APPLICATION_ENDS() {
     DEBUG_FUNCTION_LINE("Inkay: shutting down...\n");
     OSDynLoad_Release(olv_handle);
 }
+
+DECL_FUNCTION(int, FSOpenFile, FSClient *client, FSCmdBlock *block, char *path, const char *mode, uint32_t *handle, int error) {
+    const char *initialOma   = "vol/content/initial.oma";
+
+    if (strcmp(initialOma, path) == 0) {
+        //below is a hacky (yet functional!) way to get Inkay to redirect URLs from the Miiverse applet
+        //we do it when loading this file since it should only load once, preventing massive lag spikes as it searches all of MEM2 xD
```
Comment (ashquarky) [inline src/main.cpp:357]: I'm drafting a hopefully better way to handle mem2 searches, see the `devel` branch. fine to merge this for now until that's ready tho

File: src/main.cpp [hunk 6]
```diff
@@ -275,3 +348,101 @@ ON_APPLICATION_ENDS() {
...
+                DEBUG_FUNCTION_LINE("Inkay: Miiverse patches skipped.");
+            }
+        }
+    }
+    return real_FSOpenFile(client, block, path, mode, handle, error);
+}
+
+DECL_FUNCTION(FSStatus, FSReadFile, FSClient *client, FSCmdBlock *block, uint8_t *buffer, uint32_t size, uint32_t count, FSFileHandle handle, uint32_t unk1, uint32_t flags) {
+    FSStatus result = real_FSReadFile(client, block, buffer, size, count, handle, unk1, flags);
+    if(strstr((char*)buffer, "# rootca.pem") && strstr((char*)buffer, "4CE7Y259RF06alPvERck/VSyWmxzViHJbC2XpEKzJ2EFIWNt6ii8TxpvQtyYq1XT")){
```
Comment (ashquarky) [inline src/main.cpp:392]: good find, wish I knew this was here when working on Martini

**Review by ashquarky — CHANGES_REQUESTED** [review]
Looking pretty good! I'm not sure about pulling in curl but if it's a feature that we want then I can deal with it. Few small cleanup changes

**Review by ashquarky — APPROVED** [review]

**Comment by ashquarky** [comment]
I'm still kinda sad about linking libcurl, is it still possible to use the system curl lib in wut or did they scrub that? Pretendo already has lots of API endpoints set up for ancient TLS1.1-o-vision so that won't be an issue...

**Comment by ashquarky** [comment]
Merged to a work branch for further improvements:
[ ] Using libcurlwrapper from Aroma
[ ] Fixing up Dockerfile/CI
[ ] Merge conflicts (I'm sure there's one I missed)

shutter, you should have push access, let me know.

**Comment by shutterbug2000** [comment]
@ashquarky See #9 c:

**Pull request closed without merging**